openhtmltopdf-core/src/main/java/com/openhtmltopdf/util/ThreadCtx.java

```diff
@@ -2,26 +2,64 @@
 
 import com.openhtmltopdf.layout.SharedContext;
 
+import java.io.Closeable;
+import java.util.function.Consumer;
+import java.util.logging.Level;
+
 /**
  * Because OpenHTMLtoPDF is designed to run in a single thread at all times for one invocation,
  * we can use a ThreadLocal to store pseudo global variables.
  * This MUST be set up in the appropriate renderer.
  */
 public class ThreadCtx {
-	private static final ThreadLocal<ThreadData> data = new ThreadLocal<ThreadCtx.ThreadData>() {
-		@Override
-		protected ThreadData initialValue() {
-			return new ThreadData();
-		}
-	};
-	
+
+	private static final ThreadLocal<ThreadData> data = ThreadLocal.withInitial(() -> new ThreadData());
+	private static final ThreadLocal<Consumer<Diagnostic>> diagnosticConsumer = new ThreadLocal();
+
+
+
 	public static ThreadData get() {
 		return data.get();
 	}
+
+	public static void addDiagnostic(String where, Level level, String diagnostic) {
+		addDiagnostic(where, level, diagnostic, null);
+	}
+
+	public static void addDiagnostic(String where, Level level, String msg, Throwable th) {
+		Consumer<Diagnostic> consumer = diagnosticConsumer.get();
+		if (consumer != null) {
+			consumer.accept(new Diagnostic(where, level, msg, th));
+		}
+	}
 	
 	public static void cleanup() {
 		data.remove();
```

Comment:
I think we have to `diagnosticConsumer.remove()` here to avoid zombie data.

Reply:
I see on second thoughts that you have dealt with it using closeable of your own in apply method. Your method is probably cleaner.